swift/extractor/Extractor.cpp

```diff
@@ -0,0 +1,101 @@
+#include "Extractor.h"
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include <unistd.h>
+
+#include <swift/AST/SourceFile.h>
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+
+using namespace codeql;
+
+Extractor::Extractor(const Configuration& config, swift::CompilerInstance& instance)
+    : config{config}, compiler{instance} {}
+
+void Extractor::extract() {
+  if (compiler.getPrimarySourceFiles().empty()) {
+    swift::ModuleDecl* module = compiler.getMainModule();
+    if (!module->getFiles().empty() &&
+        module->getFiles().front()->getKind() == swift::FileUnitKind::Source) {
+      /// We can only call getMainSourceFile if the first file is of a Source kind
+      swift::SourceFile& file = module->getMainSourceFile();
+      extractFile(file);
+    }
+  } else {
+    for (auto s : compiler.getPrimarySourceFiles()) {
+      extractFile(*s);
+    }
+  }
+}
+
+void Extractor::extractFile(swift::SourceFile& file) {
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.trapDir)) {
+    std::cerr << "Cannot create TRAP directory: " << ec.message() << "\n";
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.sourceArchiveDir)) {
+    std::cerr << "Cannot create source archive directory: " << ec.message() << "\n";
+    return;
+  }
+
+  llvm::SmallString<PATH_MAX> srcFilePath(file.getFilename());
+  llvm::sys::fs::make_absolute(srcFilePath);
+
+  llvm::SmallString<PATH_MAX> dstFilePath(config.sourceArchiveDir);
+  llvm::sys::path::append(dstFilePath, srcFilePath);
+
+  llvm::StringRef parent = llvm::sys::path::parent_path(dstFilePath);
+  if (std::error_code ec = llvm::sys::fs::create_directories(parent)) {
+    std::cerr << "Cannot create source archive destination directory '" << parent.str()
+              << "': " << ec.message() << "\n";
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::copy_file(srcFilePath, dstFilePath)) {
+    std::cerr << "Cannot archive source file '" << srcFilePath.str().str() << "' -> '"
+              << dstFilePath.str().str() << "': " << ec.message() << "\n";
+    return;
+  }
+
+  /// TODO: find a more robust approach to avoid collisions?
+  std::string tempTrapName = file.getFilename().str() + '.' + std::to_string(getpid()) + ".trap";
```

Comment:
Why do you first create a temp file and then move it into place? This might be worthy of a comment.

Reply:
There could be several parallel processes that attempt to extract the same file.
Adding a comment is a great idea!

Reply:
Any reason we cannot use `mkstemps` function, which is robust to collisions?

Reply:
I don't fully understand the answer here: how would there be a collision with `mstemps`?

Reply:
Uh, I referred to `mktemp` which does not create the file but only gives a 'unique' file name. I take my words back 😃

Reply:
I think we can revisit this when we fix the TODO